### PR TITLE
refactor: add build system and convert to TypeScript

### PR DESCRIPTION
**Pull request by johnhooks**
This is an experiment to convert the Plaits Editor to a TypeScript project and add a build system.

The branch is a work in progress and doesn't build at the moment. I opened this pull request in order to simplify the process of inquiring about function parameter types as I work through the refactor.

## Changes:

- Move all `js` and `ts` files into `src` directory.
- Move all static files to `public` directory.
- Use build system and ES Modules to keep global scope clean rather than name spacing under global variables. 
- Use `prettier` to automatically format project files.
- Use `rollup` to create bundled modules
- Use a github workflow to build the site.
- Add the ability to connect a MIDI controller.

## Planned changes

- Use Preact or Svelte for manipulating the DOM and UI state management.
- Use WebGL shaders to render wave animations

### CODE REVIEW

File: src/encoder.ts [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+			for (const byte of block) {
+				for (let bit = 0; bit < 8; bit++) {
+					symbolStream.push(byte & (0x80 >> bit) ? 1 : 0);
+				}
+			}
+		}
+		return this.encode(new Uint8Array(symbolStream));
+	}
+
+	code(data: Uint8Array, tags: number[]) {
```
Comment (johnhooks) [inline src/encoder.ts:165]: @pichenettes I'm pretty sure this method takes in a `Uint8Array`s but I would appreciate it if you reviewed the method parameter types for this class. I think the refactor of `Encoder` is complete.

**Comment by pichenettes** [comment]
This project is published as is, as a "farewell gift" to a community I am leaving.

Unless obvious bugs are found, or significant improvements brought forward (eg: improving compatibility with other browsers or platforms), I don't intent to continue maintaining it and making it evolve – especially if it starts receiving changes in a language or environment I am not familiar with.

I encourage you to fork it.

**Comment by johnhooks** [comment]
@pichenettes I totally understand. I'll figure it out. Thanks for leaving it behind before you move on.

**Comment by johnhooks** [comment]
Closing, work will continue at https://github.com/johnhooks/plaits-editor

**Pull request closed without merging**